### ISSUE TITLE
Add Function to get UART belonging VID/DID

### ISSUE THREAD

**Issue by akaeba** [AndreRenaud/simple_uart#24]
Hello Andre,

i have the problem to select between multiple USB-TTY devices the correct. For this porpose i thought of VID/DID. With a little help from google i found https://unix.stackexchange.com/questions/144029/command-to-determine-ports-of-a-device-like-dev-ttyusb0.

As bash example:
```bash
find /sys/bus/usb/devices/usb*/ -name dev   # find usb tty devices/usb*/
find /sys/bus/serial/devices/ -name dev     # find native tty devices

udevadm info -q name -p /sys/bus/usb/devices/usb2/2-2/2-2:1.0/tty/ttyACM0       # get name
udevadm info -q property -p /sys/bus/usb/devices/usb2/2-2/2-2:1.0/tty/ttyACM0   # get VID/PID

DEVPATH=/devices/pci0000:00/0000:00:06.0/usb2/2-2/2-2:1.0/tty/ttyACM0
DEVNAME=/dev/ttyACM0
MAJOR=166
MINOR=0
SUBSYSTEM=tty
USEC_INITIALIZED=1038986785
ID_BUS=usb
ID_VENDOR_ID=04d8
ID_MODEL_ID=ffee
ID_PCI_CLASS_FROM_DATABASE=Serial bus controller
ID_PCI_SUBCLASS_FROM_DATABASE=USB controller
ID_PCI_INTERFACE_FROM_DATABASE=OHCI
ID_VENDOR_FROM_DATABASE=Apple Inc.
ID_MODEL_FROM_DATABASE=KeyLargo/Intrepid USB
ID_PATH=pci-0000:00:06.0-usb-0:2:1.0
ID_PATH_TAG=pci-0000_00_06_0-usb-0_2_1_0
ID_VENDOR=Devantech_Ltd.
ID_VENDOR_ENC=Devantech\x20Ltd.
ID_MODEL=USB-ISS.
ID_MODEL_ENC=USB-ISS.
ID_REVISION=0100
ID_SERIAL=Devantech_Ltd._USB-ISS._00060147
ID_SERIAL_SHORT=00060147
ID_TYPE=generic
ID_USB_INTERFACES=:020201:0a0000:
ID_USB_INTERFACE_NUM=00
ID_USB_DRIVER=cdc_acm
ID_MM_CANDIDATE=1
DEVLINKS=/dev/serial/by-id/usb-Devantech_Ltd._USB-ISS._00060147-if00 /dev/serial/by-path/pci-0000:00:06.0-usb-0:2:1.0
TAGS=:systemd:
```

After extracting ```ID_VENDOR_ID``` and ```ID_MODEL_ID``` i would have the chance to unambiguous indentify my USB-TTY device. 

The function prototyp could be ```int simple_uart_vid_did(const char uart[], uint32_t *vid, uint32_t *did)``` that means the VID/DID is only extracted for a single given name. This would also relax the system calls.

A similar information is under windows also available:
![image](https://github.com/AndreRenaud/simple_uart/assets/29815957/f82cc8cb-6e73-427a-adc2-e6a8b06bb792)

What do you think, should i start a try to implement the algorithm under linux?

BR,
Andeas

**Comment by akaeba** [comment]
For extracting the info from windows, i found here a link: https://www.codeguru.com/windows/determining-what-serial-ports-are-available-on-a-windows-machine/ with the related function

```c
void EnumPortsWdm(CArray<SSerInfo,SSerInfo&> &asi)
{
	CString strErr;
	// Create a device information set that will be the container for 
	// the device interfaces.
	GUID *guidDev = (GUID*) &GUID_CLASS_COMPORT;

	HDEVINFO hDevInfo = INVALID_HANDLE_VALUE;
	SP_DEVICE_INTERFACE_DETAIL_DATA *pDetData = NULL;

	try {
		hDevInfo = SetupDiGetClassDevs( guidDev,
			NULL,
			NULL,
			DIGCF_PRESENT | DIGCF_DEVICEINTERFACE
			);

		if(hDevInfo == INVALID_HANDLE_VALUE) 
		{
			strErr.Format("SetupDiGetClassDevs failed. (err=%lx)",
				GetLastError());
			throw strErr;
		}

		// Enumerate the serial ports
		BOOL bOk = TRUE;
		SP_DEVICE_INTERFACE_DATA ifcData;
		DWORD dwDetDataSize = sizeof(SP_DEVICE_INTERFACE_DETAIL_DATA) + 256;
		pDetData = (SP_DEVICE_INTERFACE_DETAIL_DATA*) new char[dwDetDataSize];
		// This is required, according to the documentation. Yes,
		// it's weird.
		ifcData.cbSize = sizeof(SP_DEVICE_INTERFACE_DATA);
		pDetData->cbSize = sizeof(SP_DEVICE_INTERFACE_DETAIL_DATA);
		for (DWORD ii=0; bOk; ii++) {
			bOk = SetupDiEnumDeviceInterfaces(hDevInfo,
				NULL, guidDev, ii, &ifcData);
			if (bOk) {
				// Got a device. Get the details.
				SP_DEVINFO_DATA devdata = {sizeof(SP_DEVINFO_DATA)};
				bOk = SetupDiGetDeviceInterfaceDetail(hDevInfo,
					&ifcData, pDetData, dwDetDataSize, NULL, &devdata);
				if (bOk) {
					CString strDevPath(pDetData->DevicePath);
					// Got a path to the device. Try to get some more info.
					TCHAR fname[256];
					TCHAR desc[256];
					BOOL bSuccess = SetupDiGetDeviceRegistryProperty(
						hDevInfo, &devdata, SPDRP_FRIENDLYNAME, NULL,
						(PBYTE)fname, sizeof(fname), NULL);
					bSuccess = bSuccess && SetupDiGetDeviceRegistryProperty(
						hDevInfo, &devdata, SPDRP_DEVICEDESC, NULL,
						(PBYTE)desc, sizeof(desc), NULL);
					BOOL bUsbDevice = FALSE;
					TCHAR locinfo[256];
					if (SetupDiGetDeviceRegistryProperty(
						hDevInfo, &devdata, SPDRP_LOCATION_INFORMATION, NULL,
						(PBYTE)locinfo, sizeof(locinfo), NULL))
					{
						// Just check the first three characters to determine
						// if the port is connected to the USB bus. This isn't
						// an infallible method; it would be better to use the
						// BUS GUID. Currently, Windows doesn't let you query
						// that though (SPDRP_BUSTYPEGUID seems to exist in
						// documentation only).
						bUsbDevice = (strncmp(locinfo, "USB", 3)==0);
					}
					if (bSuccess) {
						// Add an entry to the array
						SSerInfo si;
						si.strDevPath = strDevPath;
						si.strFriendlyName = fname;
						si.strPortDesc = desc;
						si.bUsbDevice = bUsbDevice;
						asi.Add(si);
					}

				}
				else {
					strErr.Format("SetupDiGetDeviceInterfaceDetail failed. (err=%lx)",
						GetLastError());
					throw strErr;
				}
			}
			else {
				DWORD err = GetLastError();
				if (err != ERROR_NO_MORE_ITEMS) {
					strErr.Format("SetupDiEnumDeviceInterfaces failed. (err=%lx)", err);
					throw strErr;
				}
			}
		}
	}
	catch (CString strCatchErr) {
		strErr = strCatchErr;
	}

	if (pDetData != NULL)
		delete [] (char*)pDetData;
	if (hDevInfo != INVALID_HANDLE_VALUE)
		SetupDiDestroyDeviceInfoList(hDevInfo);

	if (!strErr.IsEmpty())
		throw strErr;
}
```

I tryed also to play a little with chat gpt.

**Comment by AndreRenaud** [comment]
This would be reasonable. That example code is for C++, and uses exceptions which wouldn't be suitable (the library is currently straight C). But it could be reworked into straight C. I did have the idea of having a string description for each port, which could include `vid=XYZ,pid=XYZ` kind of details, and if we had that, then having a function for opening by grepping through these strings would be pretty straight forward (much like the `hwgrep` feature of pyserial - https://pyserial.readthedocs.io/en/latest/url_handlers.html#hwgrep)

**Comment by andkae** [comment]
Hi Andre,

> But it could be reworked into straight C.
This i can try. And add an straight forward C code as comment here.


> I did have the idea of having a string description for each port, which could include `vid=XYZ,pid=XYZ`

We should try to define as first the interface and the name of the function. For instance: ```int simple_uart_prop(const char uart[], char ***propesp)``` and then the info to add. Because under windows it could be easily became a mess to collect all the infos.

BR,
Andreas

**Comment by andkae** [comment]
In relation to pyserial it could also named ```int simple_uart_hwgrep(const char uart[], char ***propesp)```

**Comment by AndreRenaud** [comment]
I think there's probably two functions. The one that gives the info, which could be something like:
`int simple_uart_describe(const char *uart_device, char *description, int max_description_length)`

And then a new function for automatically opening the first matching device, ie:
`struct simple_uart *simple_uart_open_match(const char *match, int speed, const char *mode_string)`
Which runs through all the available ports finding one that contains `match` as a substring of its description, and then tries to open the first matching device.

**Comment by andkae** [comment]
Hi André

yes this function ```int simple_uart_describe(const char *uart_device, char *description, int max_description_length)``` sounds reasonable. But we take into account how should this function behaves in windows and linux. Do we define a set of keywords, like you mention in:
> vid=XYZ,pid=XYZ
this is then a reformat of the ```udevadm info -q property -p``` output, or the functionoutputs only the implemented info which could gives a lot of cnfusion between windows/linux environemnts. From my point of view i would prefer a own key definition which is then equal between linux and windows. For my case would the ```vid/did``` enough and later can this function mature. But please equal between linux/windows.


> struct simple_uart *simple_uart_open_match(const char *match, int speed, const char *mode_string)
Also d'accord, but please keep the old function as well:
```struct simple_uart *simple_uart_open(const char *device, int speed, const char *mode_string)```

Some doubts two the ```*match``` string:
* What happends if the user like to check the ```vid``` and a future match criteria. At the end of the day you have to build an key/value pair list between criterias to match and the read criteria. Perhaps would it better to output a key value list, i think it could be easier to match, instead of grepping in two strings and compare them.

What do you think?

BR,
Andreas

**Comment by andkae** [comment]
My function proposal would look like follows:

```int simple_uart_describe(const char *uart_device, char ***keyp, char ***valp)``` For the return:

| return | meaning |
| < 0     | ero |
| >= 0  | number of returned keys |

Could this also work?

BR,
Andreas

**Comment by akaeba** [comment]
Hi Andre,

i started right now with the uart_describe implementation, see #31. The Win32 part is in my opinion ready.

BR,
Andreas

**Comment by andkae** [comment]
solved in #31 Thanks for merge.
